extmod/modbluetooth.c

```diff
@@ -500,24 +500,26 @@ STATIC mp_obj_t bluetooth_ble_gap_connect(size_t n_args, const mp_obj_t *args) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(bluetooth_ble_gap_connect_obj, 3, 4, bluetooth_ble_gap_connect);
 
-STATIC mp_obj_t bluetooth_ble_gap_scan(size_t n_args, const mp_obj_t *args) {
-    if (n_args == 2 && args[1] == mp_const_none) {
-        int err = mp_bluetooth_gap_scan_stop();
-        return bluetooth_handle_errno(err);
-    } else {
-        mp_int_t duration_ms = 0;
-        if (n_args == 2) {
-            if (!mp_obj_is_int(args[1])) {
-                mp_raise_ValueError("invalid duration");
-            }
-            duration_ms = mp_obj_get_int(args[1]);
-        }
-
-        int err = mp_bluetooth_gap_scan_start(duration_ms);
-        return bluetooth_handle_errno(err);
+STATIC mp_obj_t bluetooth_ble_gap_scan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_duration_ms, ARG_interval_ms, ARG_window_ms, ARG_active };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_duration_ms, MP_ARG_OBJ, {.u_obj = MP_OBJ_NEW_SMALL_INT(0)} },
+        { MP_QSTR_interval_ms, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_window_ms, MP_ARG_INT, {.u_int = 0} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
```

Comment:
Instead of kwargs it can just use positional with defaults.  That should save code size, no need for qstrs nor the table of allowed args.

And I think the default values should be specified, eg to be what they currently are.  Something like:
```python
mp_int_t interval_ms = 1280;
mp_int_t window_ms = 11;
if (n_args > 2) {
    interval_ms = mp_obj_get_int(args[2]);
    if (n_args > 3) {
        window_ms = mp_obj_get_int(args[3]);
    }
}
```

Reply:
Oh, I see that `gap_advertise` uses kwargs.  Nevertheless I still think it makes sense to keep `gap_scan` "simple" and just use positional here.